src/module.c

```diff
@@ -10987,10 +11122,384 @@ sds genModulesInfoString(sds info) {
+    }
+    if (type != NUMERIC_CONFIG && flags & REDISMODULE_CONFIG_MEMORY) {
+        return REDISMODULE_ERR;
+    } else if (type == NUMERIC_CONFIG) {
+        int numeric_flag = flags & (REDISMODULE_CONFIG_MEMORY);
+        /* Can only provide one numeric flag. */
+        if (!(numeric_flag == REDISMODULE_CONFIG_MEMORY || !numeric_flag))
+            return REDISMODULE_ERR;
+    }
+    return REDISMODULE_OK;
+}
+
+/* This is a series of set functions for each type that act as dispatchers for 
+ * config.c to call module set callbacks. */
+int setModuleBoolConfig(ModuleConfig *config, int val, const char **err) {
+    int return_code = config->set_fn.set_bool(config->name, val, config->privdata, err);
+    return return_code == REDISMODULE_OK ? 1 : 0;
+}
+
+int setModuleStringConfig(ModuleConfig *config, sds strval, const char **err) {
+    RedisModuleString *prev = config->get_fn.get_string(config->name, config->privdata);
+    RedisModuleString *new = createStringObject(strval, sdslen(strval));
+    int return_code = config->set_fn.set_string(config->name, new, config->privdata, err);
+    if (prev) decrRefCount(prev);
+    decrRefCount(new);
+    return return_code == REDISMODULE_OK ? 1 : 0;
+}
+
+int setModuleEnumConfig(ModuleConfig *config, int val, const char **err) {
+    int return_code = config->set_fn.set_enum(config->name, val, config->privdata, err);
+    return return_code == REDISMODULE_OK ? 1 : 0;
+}
+
+int setModuleNumericConfig(ModuleConfig *config, long long val, const char **err) {
+    int return_code = config->set_fn.set_numeric(config->name, val, config->privdata, err);
+    return return_code == REDISMODULE_OK ? 1 : 0;
+}
+
+/* This is a series of get functions for each type that act as dispatchers for 
+ * config.c to call module set callbacks. */
+int getModuleBoolConfig(ModuleConfig *module_config) {
+    return module_config->get_fn.get_bool(module_config->name, module_config->privdata);
+}
+
+sds getModuleStringConfig(ModuleConfig *module_config) {
+    RedisModuleString *val = module_config->get_fn.get_string(module_config->name, module_config->privdata);
+    return val ? sdsdup(val->ptr) : NULL;
+}
+
+int getModuleEnumConfig(ModuleConfig *module_config) {
+    return module_config->get_fn.get_enum(module_config->name, module_config->privdata);
+}
+
+long long getModuleNumericConfig(ModuleConfig *module_config) {
+    return module_config->get_fn.get_numeric(module_config->name, module_config->privdata);
+}
+
+/* This function takes a module and a list of configs stored as sds NAME VALUE pairs.
+ * It attempts to call set on each of these configs. */
+int loadModuleConfigs(RedisModule *module) {
+    if (!server.module_configs_queue || !listLength(server.module_configs_queue)) return REDISMODULE_OK;
+    listIter li;
+    listNode *ln;
+    int count;
+    sds *config_pair;
+    const char *err = NULL;
+
+    listRewind(server.module_configs_queue, &li);
+    while ((ln = listNext(&li))) {
```

Comment:
i think this is the place where we're also suppose to apply the default values to the module.
i.e. call the `setfn` callback with either the value provided by the user, or the default value.
if we don't call the setter with the default value on load time, we can get a case where there's a mismatch between the default value for execution, and the default value for rewrite.

to do that, i suppose we need to change the loop to iterate over the registered configs, rather than the ones provided by the user (module_configs_queue).